## Deriving New Variables

## Following Steps:

### Step 1(a). Creating a “price_label” column.

### Step 1(b). Creating "busiest_day" column.

### Step 2. Updating "busiest day" column to "busiest days" to identify order day of the week as "Busiest days," "Slowest days," or "Regularly busy."


### Step 3. Checking the values of new "busiest days" column for accuracy with observations in markdown format.


### Step 4. Creating a new column "busiest_period_of_day" to identify time periods “Most orders,” “Average orders,” and “Fewest orders.”


### Step 5. Print the frequency of "busiest_period_of_day" column.


### Step 6. Export dataframe as a pickle file to “Prepared Data” folder.

### Import libraries

In [106]:
# Import libraries

import pandas as pd
import numpy as np
import os

### Import the data sets into Jupyter

In [22]:
# Tell Python to remember a main folder path
path = r'/Users/ashwanisherawat/Downloads/Instacart Basket Analysis/02 Data'

In [30]:
# Import dataset orders_products_combined.pkl
df_ords_prods_merge = pd.read_pickle(os.path.join(path, 'Prepared Data', 'ords_prods_merge.pkl'))

### Step 1(a). Creating a “price_label” column.

In [47]:
# Create a subset of first million rows

df = df_ords_prods_merge[:1000000]

In [48]:
df_ords_prods_merge.shape

(32435059, 15)

In [49]:
df_ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     NaN         196                  1          0   both   
1                     NaN       14084                  2          0   both   
2                     NaN       12427                  3          0   both   
3                     NaN       26088                  4          0   both   
4                     NaN       26405                  5          0   both   

   Unnamed: 0                             product_name  aisle_id  \
0       195.0                                     Soda      77.0   
1     14084.0  Organic Unsweetened Vanilla Almond Milk      91.0   
2     12427.0                      Original Beef Jerky      23.0   
3     26089.0               Aged White Cheddar Popcorn      23.0   
4     26406.0         XL Pick-A-Size Paper Towel Rolls      54.0   

   department_id  prices  
0            7.0     9.0  
1           16.0    12.5  
2           19.0     4.4  
3           19.0     4.7  
4           17.0     1.0

In [66]:
# Define a function for price labels

def price_label(row):
    if row['prices'] <= 5:
        return 'Low-range product'
    elif (row['prices'] > 5) and (row['prices'] <= 15):
        return 'Mid-range product'
    elif row['prices'] > 15:
        return 'High range'
    else: return 'Not enough data'

In [67]:
# Apply price range function. Note to self: Python can technically execute this, but it will suggest the loc() method instead.

df['price_range'] = df.apply(price_label, axis=1)

/var/folders/bv/0q7zxbg57q32000r2xxtsc4c0000gn/T/ipykernel_16145/774635379.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [69]:
df['price_range'].value_counts(dropna = False)

Mid-range product    672525
Low-range product    314109
High range            12413
Not enough data         953
Name: price_range, dtype: int64

In [72]:
# Find most expensive product.

df_ords_prods_merge['prices'].max()

99999.0

In [73]:
# Remove rows where price_range is 'High range' or 'Not enough data'
df_filtered = df[~df['price_range'].isin(['High range', 'Not enough data'])]

# Verify the updated counts
print(df_filtered['price_range'].value_counts())


Mid-range product    672525
Low-range product    314109
Name: price_range, dtype: int64


### Comments 
#### - Removed the High range and Not enogh data groups form the data frame as it was not fitting into the requirements. 
#### - The analysis reveals that most products fall into the mid-range price category (672,525), followed by low-range products (314,109). This indicates a stronger consumer preference or demand for mid-priced products,highlighting potential opportunities to focus on this segment for marketing and inventory strategies.

In [84]:
# Find most expensive product.

df_filtered['prices'].max()

15.0

In [86]:
# Use loc() method on df subset to apply price range function.

df_filtered.loc[df_filtered['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [87]:
df_filtered.loc[(df_filtered['prices'] <= 15) & (df_filtered['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [88]:
df_filtered.loc[df_filtered['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [90]:
df_filtered['price_range_loc'].value_counts(dropna = False)

Mid-range product    672525
Low-range product    314109
Name: price_range_loc, dtype: int64

### Step 1(b). Creating "busiest_day" column.

In [91]:
# Change "order_dow" column name to "order_day_of_week."

df_ords_prods_merge.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

In [92]:
# Find frequency of orders_day_of_week.

df_ords_prods_merge['orders_day_of_week'].value_counts(dropna = False)

0    6209808
1    5665951
6    4500391
2    4217868
5    4209603
3    3844175
4    3787263
Name: orders_day_of_week, dtype: int64

In [93]:
# Use for-loop to determine whether orders are on "busiest day" (0 = Saturday), "least busy" (4 = Wednesday), or "regularly busy" (other days of the week).

result = []

for value in df_ords_prods_merge["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [94]:
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy

In [95]:
# Create new column in ords_prods_merge to view results in context.

df_ords_prods_merge['busiest_day'] = result

In [96]:
df_ords_prods_merge['busiest_day'].value_counts(dropna = False)

Regularly busy    22437988
Busiest day        6209808
Least busy         3787263
Name: busiest_day, dtype: int64

In [98]:
# Check output of ords_prods_merge with new "busiest days" column.

df_ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     NaN         196                  1          0   both   
1                     NaN       14084                  2          0   both   
2                     NaN       12427                  3          0   both   
3                     NaN       26088                  4          0   both   
4                     NaN       26405                  5          0   both   

   Unnamed: 0                             product_name  aisle_id  \
0       195.0                                     Soda      77.0   
1     14084.0  Organic Unsweetened Vanilla Almond Milk      91.0   
2     12427.0                      Original Beef Jerky      23.0   
3     26089.0               Aged White Cheddar Popcorn      23.0   
4     26406.0         XL Pick-A-Size Paper Towel Rolls      54.0   

   department_id  prices     busiest_day  
0            7.0     9.0  Regularly busy  
1           16.0    12.5  Regularly busy  
2           19.0     4.4  Regularly busy  
3           19.0     4.7  Regularly busy  
4           17.0     1.0  Regularly busy

### Step 2. Updating "busiest day" column to "busiest days" to identify order day of the week as "Busiest days," "Slowest days," or "Regularly busy."


In [107]:
# Create column “busiest_days" from "orders_day_of_week" to identify two busiest days (0, 1), slowest days (4, 3), or regularly busy (all other days).

result_2 = []

for value in df_ords_prods_merge["orders_day_of_week"]:
  if value == 0 or value == 1:
    result_2.append("Busiest days")
  elif value == 4 or value == 3:
    result_2.append("Slowest days")
  else:
    result_2.append("Regularly busy")

In [108]:
result_2

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest day

### Step 3. Checking the values of new "busiest days" column for accuracy with observations in markdown format.


In [109]:
# Create new 'busiest_days' column in ords_prods_merge to view results in context.

df_ords_prods_merge['busiest_days'] = result_2

In [110]:
df_ords_prods_merge['busiest_days'].value_counts(dropna = False)

Regularly busy    12927862
Busiest days      11875759
Slowest days       7631438
Name: busiest_days, dtype: int64

In [112]:
# Check output of ords_prods_merge with new "busiest days" column.

df_ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     NaN         196                  1          0   both   
1                     NaN       14084                  2          0   both   
2                     NaN       12427                  3          0   both   
3                     NaN       26088                  4          0   both   
4                     NaN       26405                  5          0   both   

   Unnamed: 0                             product_name  aisle_id  \
0       195.0                                     Soda      77.0   
1     14084.0  Organic Unsweetened Vanilla Almond Milk      91.0   
2     12427.0                      Original Beef Jerky      23.0   
3     26089.0               Aged White Cheddar Popcorn      23.0   
4     26406.0         XL Pick-A-Size Paper Towel Rolls      54.0   

   department_id  prices     busiest_day busiest_period_of_day    busiest_days  
0            7.0     9.0  Regularly busy        Average orders  Regularly busy  
1           16.0    12.5  Regularly busy        Average orders  Regularly busy  
2           19.0     4.4  Regularly busy        Average orders  Regularly busy  
3           19.0     4.7  Regularly busy        Average orders  Regularly busy  
4           17.0     1.0  Regularly busy        Average orders  Regularly busy

## Comments

The analysis shows that most orders occur on regularly busy days (12.9M), followed by the busiest days (11.8M), with the least on slowest days (7.6M). This highlights predictable shopping patterns, suggesting opportunities to optimize resources during peak times and boost engagement during slower periods.

### Step 4. Creating a new column "busiest_period_of_day" to identify time periods “Most orders,” “Average orders,” and “Fewest orders.”


In [113]:
# Check value counts in "order_hour_of_day" column.

df_ords_prods_merge['order_hour_of_day'].value_counts()

10    2764476
11    2738647
14    2691598
15    2664583
13    2663346
12    2620898
16    2537506
9     2456751
17    2089510
8     1719991
18    1637956
19    1259416
20     977049
7      891951
21     796379
22     634743
23     402621
6      290796
0      218951
1      115787
5       88064
2       69435
4       53284
3       51321
Name: order_hour_of_day, dtype: int64

## Comments
The value counts listed above will be split into equal thirds for the following labels: "Most orders" = 10, 11, 14, 15, 13, 12, 16, and 9; "Fewest orders" = 23, 6, 0, 1, 5, 2, 4, and 3; "Average orders" = all remaining values in order_hour_of_day.

In [100]:
# Create for-loop if statement labeling periods of time as “Most orders,” “Average orders,” and “Fewest orders.”

result_3 = []

for value in df_ords_prods_merge["order_hour_of_day"]:
  if value in [10, 11, 14, 15, 13, 12, 16, 9]:
    result_3.append("Most orders")
  elif value in [23, 6, 0, 1, 5, 2, 4, 3]:
    result_3.append("Fewest orders")
  else:
    result_3.append("Average orders")

In [101]:
result_3

['Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Avera

In [ ]:
# Create new column "busiest_period_of_day" in ords_prods_merge.

df_ords_prods_merge['busiest_period_of_day'] = result_3

## Step 5. Print frequency of "busiest_period_of_day" column.


In [103]:
# Print value counts in "busiest_period_of_day" column.

df_ords_prods_merge['busiest_period_of_day'].value_counts(dropna = False)

Most orders       21137805
Average orders    10006995
Fewest orders      1290259
Name: busiest_period_of_day, dtype: int64

### Step 6. Export dataframe as a pickle file to “Prepared Data” folder.

In [105]:
df_ords_prods_merge.to_pickle(os.path.join(path, 'Prepared Data', 'orders_products_merged_derived.pkl'))